string_manipulation.rb

```diff
@@ -1,30 +1,104 @@
 # A method to reverse a string in place.
 def string_reverse(my_string)
-  puts "NOT IMPLEMENTED"
+  # swap each element
+  # move an index
+  i = 0
+  n = my_string.length
+  while i < n/2
+    first = my_string[i]
+    my_string[i] = my_string[n-i-1]
+    my_string[n-i-1] = first
+    i += 1
+  end
 end
 
 # A method to reverse each word in a sentence, in place.
 def reverse_words(my_words)
-  puts "NOT IMPLEMENTED"
+  # iterate through each characters
+  # if space leave alone
+  # create stack of letters
+  # pop off onto string
+  n = my_words.length
+  i = 0
+  # CAN_ 0123 NAC
+  while i < n
+    word = []
+    j = i
+    while my_words[j] != " " && j < n
+      word << my_words[j]
+      j += 1
+    end
+    while i < j
+      my_words[i] = word.pop
+      i += 1
+    end
+    i += 1
+  end

```

Comment:
In it's current implementation, this is a O(n^2) time and O(n) space complexity algorithm. The temporary/auxiliary storage, word will be dependent on input size. If there is only one word, all the n characters will get copied over to word and then back to my_words in the correct order. This makes the space complexity O(n).
Can you think of a way to make this an O(n^2) time and O(1) space complexity? Hint: Consider creating a helper method that reverses a part of a string in place given a starting index and an ending index. Then call this helper method passing in where a word begins and where a word ends to it.